In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sns

sns.set_theme()

# jupyter notebook full-width display
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

# no text wrapping
display(HTML("<style>.dataframe td { white-space: nowrap; }</style>"))

# pandas formatting
pd.set_option('display.float_format', '{:.3f}'.format)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 200)
pd.set_option('display.max_colwidth', 400)

In [2]:
df_SD = pd.read_pickle(r'.\intermediate_data_and_calcs\df_SD.pickle')
df_FD = pd.read_pickle(r'.\intermediate_data_and_calcs\df_FD.pickle')
df_LF = pd.read_pickle(r'.\intermediate_data_and_calcs\df_LF.pickle')
df_LF_grouped = pd.read_pickle(r'.\intermediate_data_and_calcs\df_LF_grouped.pickle')
df_Site = pd.read_pickle(r'.\intermediate_data_and_calcs\df_Site.pickle')
df_TrapSupervisors = pd.read_pickle(r'.\intermediate_data_and_calcs\df_TrapSupervisors.pickle')

# checking stuff when questions come up

### why don't these match?
* answer: they do, but there is no matching sample to link them

In [3]:
df_LF[(df_LF.yy==2019) & (df_LF.mm==5) & (df_LF.dd==28) & (df_LF.site=="47")]

yy  mm  dd  Time     river  week site    loc period  wt_lbs  wt_kg  \
11195  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11196  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11197  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11198  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11199  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11200  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11201  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11202  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11203  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11204  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11205  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   

       lgth  freq  Flbin   DATETIME  SITE1 SITE2 SITE3          id  \
11195   215     1    215 2019-05-28     47  <NA>  <NA>  2019052847   
11196   220     6    220 2019-05-28     47  <NA>  <NA>  2019052847   
11197   225    16    225 2019-05-28     47  <NA>  <NA>  2019052847   
11198   230    26    230 2019-05-28     47  <NA>  <NA>  2019052847   
11199   235    39    235 2019-05-28     47  <NA>  <NA>  2019052847   
11200   240    51    240 2019-05-28     47  <NA>  <NA>  2019052847   
11201   245    35    245 2019-05-28     47  <NA>  <NA>  2019052847   
11202   250    27    250 2019-05-28     47  <NA>  <NA>  2019052847   
11203   255    17    255 2019-05-28     47  <NA>  <NA>  2019052847   
11204   260     5    260 2019-05-28     47  <NA>  <NA>  2019052847   
11205   265     1    265 2019-05-28     47  <NA>  <NA>  2019052847   

       length_bin_id FLAG_SITE FLAG_AM_PM_PERIOD  \
11195         21.500       NaN               NaN   
11196         22.000       NaN               NaN   
11197         22.500       NaN               NaN   
11198         23.000       NaN               NaN   
11199         23.500       NaN               NaN   
11200         24.000       NaN               NaN   
11201         24.500       NaN               NaN   
11202         25.000       NaN               NaN   
11203         25.500       NaN               NaN   
11204         26.000       NaN               NaN   
11205         26.500       NaN               NaN   

      FLAG_AM_PM_PERIOD_DISCREPANCIES FLAG_NO_MATCHING_SAMPLE  
11195                             NaN                    True  
11196                             NaN                    True  
11197                             NaN                    True  
11198                             NaN                    True  
11199                             NaN                    True  
11200                             NaN                    True  
11201                             NaN                    True  
11202                             NaN                    True  
11203                             NaN                    True  
11204                             NaN                    True  
11205                             NaN                    True

In [4]:
df_SD[(df_SD.DATETIME.dt.year==2019) & (df_SD.DATETIME.dt.month==5) & (df_SD.DATETIME.dt.day==28)]

DIST   RIVER                 NAME code  GEAR SITE_NO no_nets  YEAR  MM  \
14946     2  SWMARG     Martin E Cameron  NaN    81      12       1  2019   5   
14947     2  SWMARG        Robert Peters  NaN    81      25       1  2019   5   
14948     2  SWMARG      Pierre Chiasson  NaN    81      26       1  2019   5   
14949     2  SWMARG    Gerard MacFarlane  NaN    81      28       1  2019   5   
14950     2  SWMARG       Daniel Stewart  NaN    81      33       1  2019   5   
14951     2  SWMARG       Stewart Gillis  NaN    81      41       1  2019   5   
14952     2  SWMARG       Finley Stewart  NaN    81      44     NaN  2019   5   
14953     2  SWMARG        Mary E Gillis  NaN    81      49       1  2019   5   
14954     2  SWMARG      Bruce MacLellan  NaN    81      51       1  2019   5   
14955     2  SWMARG  Elizabeth MacKinnon  NaN    81      62       1  2019   5   
14956     2  SWMARG      Kevin MacKinnon  NaN    81      64       1  2019   5   

       DD  Week  catch_lbs  catch_kg hours_fished   zone   last_name comments  \
14946  28     5    350.000   158.800            5  lower     Cameron      NaN   
14947  28     5     25.000    11.300            8  lower      Peters      NaN   
14948  28     5    200.000    90.700            7  lower    Chiasson      NaN   
14949  28     5    200.000    90.700            8  upper  MacFarlane      NaN   
14950  28     5    150.000    68.000            8  upper     Stewart      NaN   
14951  28     5     30.000    13.600            7  upper      Gillis      NaN   
14952  28     5    300.000   136.100           14  upper     Stewart      NaN   
14953  28     5    700.000   317.500           14  upper      Gillis      NaN   
14954  28     5     15.000     6.800            6  upper   MacLellan      NaN   
14955  28     5    500.000   226.800           10  upper   MacKinnon      NaN   
14956  28     5    120.000    54.400            6  upper   Mackinnon      NaN   

       bycatch_sbass  bycatch_shad bycatch_other   DATETIME  SITE1 SITE2  \
14946            NaN           NaN           NaN 2019-05-28     12  <NA>   
14947            NaN           NaN           NaN 2019-05-28     25  <NA>   
14948            NaN           NaN           NaN 2019-05-28     26  <NA>   
14949            NaN           NaN       1 perch 2019-05-28     28  <NA>   
14950            NaN           NaN           NaN 2019-05-28     33  <NA>   
14951            NaN           NaN           NaN 2019-05-28     41  <NA>   
14952            NaN           NaN           NaN 2019-05-28     44  <NA>   
14953            NaN           NaN           NaN 2019-05-28     49  <NA>   
14954            NaN           NaN           NaN 2019-05-28     51  <NA>   
14955            NaN           NaN           NaN 2019-05-28     62  <NA>   
14956            NaN           NaN           NaN 2019-05-28     64  <NA>   

                      remarks          id  total_fish_preserved  \
14946                     NaN  2019052812                31.000   
14947                     NaN  2019052825                   NaN   
14948                     NaN  2019052826                   NaN   
14949  bycatch_other: 1 perch  2019052828                   NaN   
14950                     NaN  2019052833                   NaN   
14951                     NaN  2019052841                   NaN   
14952                     NaN  2019052844                   NaN   
14953                     NaN  2019052849                   NaN   
14954                     NaN  2019052851                   NaN   
14955                     NaN  2019052862                36.000   
14956                     NaN  2019052864                   NaN   

       total_fish_measured AM_PM_PERIOD  wt_lbs FLAG_DATETIME  \
14946              195.000           AM  71.900           NaN   
14947                  NaN          NaN     NaN           NaN   
14948                  NaN          NaN     NaN           NaN   
14949                  NaN          NaN     NaN           NaN   
14950             

### kevin followup question: how many ambiguous matches don't have a matching sample because there are multiple sample matches

In [5]:
ambiguous = set()
for i in list(df_SD[df_SD.id>2024000000].id):
    while i > 2024000000:
        i -= 1000000000
    ambiguous.add(i)
    
ambiguous = list(ambiguous)

In [6]:
# EXACT MATCHES = NONE

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # no exact matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY) & (df_LF.site==str(SITE))])
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY) & (df_FD.SITE==SITE)])


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

In [7]:
# POTENTIAL MATCHES WITH MESSY SITE NAMES = NONE

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # partial matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY) & (df_LF.site.str.contains(str(SITE))==True)])
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY) & (df_FD.SITE.str.contains(str(SITE))==True)])


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

In [8]:
# A FEW MATCHES WHEN SITE IS NOT CONSIDERED
# maybe there the site number was written down wrong, some partial matches (excluding site)

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # very partial matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY)].head())
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY)].head())


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


yy  mm  dd  Time        river  week site    loc period  wt_lbs  wt_kg  \
3024  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3025  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3026  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3027  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3028  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   

      lgth  freq  Flbin   DATETIME  SITE1 SITE2 SITE3          id  \
3024   220     3    220 1997-06-12     26  <NA>  <NA>  1997061226   
3025   225     4    225 1997-06-12     26  <NA>  <NA>  1997061226   
3026   230    19    230 1997-06-12     26  <NA>  <NA>  1997061226   
3027   235    52    235 1997-06-12     26  <NA>  <NA>  1997061226   
3028   240    70    240 1997-06-12     26  <NA>  <NA>  1997061226   

      length_bin_id FLAG_SITE FLAG_AM_PM_PERIOD  \
3024         22.000       NaN               NaN   
3025         22.500       NaN               NaN   
3026         23.000       NaN               NaN   
3027         23.500       NaN               NaN   
3028         24.000       NaN               NaN   

     FLAG_AM_PM_PERIOD_DISCREPANCIES FLAG_NO_MATCHING_SAMPLE  
3024                             NaN                     NaN  
3025                             NaN                     NaN  
3026                             NaN                     NaN  
3027                             NaN                     NaN  
3028                             NaN                     NaN

YEAR  MM  DD  WEEK SITE PERIOD CONDITION  FISH_NO  FL_WET  FL_FROZEN  \
15290  1997   6  12     7   26     AM    Frozen        1    <NA>        230   
15291  1997   6  12     7   26     AM    Frozen        2    <NA>        224   
15292  1997   6  12     7   26     AM    Frozen        3    <NA>        272   
15293  1997   6  12     7   26     AM    Frozen        4    <NA>        212   
15294  1997   6  12     7   26     AM    Frozen        5    <NA>        220   

       FL_STD  WEIGHT SPECIES SEX  MATURITY  GONAD_WEIGHT Ager_1  AGE_1  \
15290 238.000 171.000       A   M         4           NaN    NaN      3   
15291 232.000 166.000       A   M         4           NaN    NaN      3   
15292 280.000 307.000       B   F         4        39.900    NaN      7   
15293 220.000 138.000       A   M         4           NaN    NaN      3   
15294 228.000 143.000       A   M         4           NaN    NaN      3   

       FSP_1 Comments_1 Ager_2  AGE_2  FSP_2 Comments_2 Ager_3  AGE_3  FSP_3  \
15290      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15291      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15292      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15293      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15294      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   

      Comments_3 Envelop.Comments   DATETIME AGE_notes_1 FSP_notes_1  \
15290        NaN              NaN 1997-06-12           3           3   
15291        NaN              NaN 1997-06-12           3           3   
15292        NaN              NaN 1997-06-12           7           3   
15293        NaN              NaN 1997-06-12           3           3   
15294        NaN              NaN 1997-06-12           3           3   

      AGE_notes_2 FSP_notes_2 AGE_notes_3 FSP_notes_3 SITE_notes  SITE1 SITE2  \
15290         NaN         NaN         NaN         NaN         26     26  <NA>   
15291         NaN         NaN         NaN         NaN         26     26  <NA>   
15292         NaN         NaN         NaN         NaN         26     26  <NA>   
15293         NaN         NaN         NaN         NaN         26     26  <NA>   
15294         NaN         NaN         NaN         NaN         26     26  <NA>   

      SITE3  fish_length                                         remarks  \
15290  <NA>          230  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15291  <NA>          224  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15292  <NA>          272  SITE_notes: 26; AGE_notes_1: 7; FSP_notes_1: 3   
15293  <NA>          212  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15294  <NA>          220  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   

               id FLAG_SITE FLAG_AM_PM_PERIOD FLAG_SEX FLAG_MATURITY  \
15290  1997061226       NaN               NaN      NaN           NaN   
15291  1997061226       NaN               NaN      NaN           NaN   
15292  1997061226       NaN               NaN      NaN           NaN   
15293  1997061226       NaN               NaN      NaN           NaN   
15294  1997061226       NaN               NaN      NaN           NaN   

      FLAG_FSP_1 FLAG_FL_STD FLAG_FL_WET_FROZEN FLAG_WEIGHT_OUTLIER  \
15290        NaN         NaN                NaN                 NaN   
15291        NaN         NaN                NaN                 NaN   
15292        NaN         NaN                NaN                 NaN   
15293        NaN         NaN                NaN                 NaN   
15294        NaN         NaN                NaN                 NaN   

      FLAG_GONAD_OUTLIER FLAG_MULTIPLE_SAMPLE_POSSIBILITIES  \
15290                NaN                                NaN   
15291                NaN                                NaN   
15292                NaN                                NaN   
15293                NaN                                NaN   
15294                NaN                        


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

YEAR  MM  DD  WEEK SITE PERIOD CONDITION  FISH_NO  FL_WET  FL_FROZEN  \
4503  1988   5  23  <NA>   12    NaN     Fresh        1     253       <NA>   
4504  1988   5  23  <NA>   12    NaN     Fresh        2     258       <NA>   
4505  1988   5  23  <NA>   12    NaN     Fresh        3     245       <NA>   
4506  1988   5  23  <NA>   12    NaN     Fresh        4     250       <NA>   
4507  1988   5  23  <NA>   12    NaN     Fresh        5     245       <NA>   

      FL_STD  WEIGHT SPECIES SEX  MATURITY  GONAD_WEIGHT Ager_1  AGE_1  FSP_1  \
4503 253.000 208.000       A   M         4           NaN    NaN      4      4   
4504 258.000 236.000       A   M         4           NaN    NaN      4      4   
4505 245.000     NaN       A   M         4           NaN    NaN      4      4   
4506 250.000 234.000       A   M         4           NaN    NaN      4      4   
4507 245.000 209.000       A   F         4        29.100    NaN      3      3   

     Comments_1 Ager_2  AGE_2  FSP_2 Comments_2 Ager_3  AGE_3  FSP_3  \
4503        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4504        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4505        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4506        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4507        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   

     Comments_3 Envelop.Comments   DATETIME AGE_notes_1 FSP_notes_1  \
4503        NaN              NaN 1988-05-23           4           4   
4504        NaN              NaN 1988-05-23           4           4   
4505        NaN              NaN 1988-05-23           4           4   
4506        NaN              NaN 1988-05-23           4           4   
4507        NaN              NaN 1988-05-23           3           3   

     AGE_notes_2 FSP_notes_2 AGE_notes_3 FSP_notes_3 SITE_notes  SITE1 SITE2  \
4503         NaN         NaN         NaN         NaN         12     12  <NA>   
4504         NaN         NaN         NaN         NaN         12     12  <NA>   
4505         NaN         NaN         NaN         NaN         12     12  <NA>   
4506         NaN         NaN         NaN         NaN         12     12  <NA>   
4507         NaN         NaN         NaN         NaN         12     12  <NA>   

     SITE3  fish_length                                         remarks  \
4503  <NA>          253  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4504  <NA>          258  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4505  <NA>          245  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4506  <NA>          250  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4507  <NA>          245  SITE_notes: 12; AGE_notes_1: 3; FSP_notes_1: 3   

              id FLAG_SITE FLAG_AM_PM_PERIOD FLAG_SEX FLAG_MATURITY  \
4503  1988052312       NaN               NaN      NaN           NaN   
4504  1988052312       NaN               NaN      NaN           NaN   
4505  1988052312       NaN               NaN      NaN           NaN   
4506  1988052312       NaN               NaN      NaN           NaN   
4507  1988052312       NaN               NaN      NaN           NaN   

     FLAG_FSP_1 FLAG_FL_STD FLAG_FL_WET_FROZEN FLAG_WEIGHT_OUTLIER  \
4503        NaN         NaN                NaN                 NaN   
4504        NaN         NaN                NaN                 NaN   
4505        NaN         NaN                NaN                 NaN   
4506        NaN         NaN                NaN                 NaN   
4507        NaN         NaN                NaN                 NaN   

     FLAG_GONAD_OUTLIER FLAG_MULTIPLE_SAMPLE_POSSIBILITIES  \
4503                NaN                                NaN   
4504                NaN                                NaN   
4505                NaN                                NaN   
4506                NaN                                NaN   
4507                NaN                                NaN   

     FLAG_MISNUMBERED_FISH_DETAILS FL


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

### how do we flag and leave the potential to merge ghost/ambiguous samples?

In [9]:
id = 1990061312
date = id // 100
date

19900613

In [10]:
from datetime import datetime
datetime.strptime(str(id//100), "%Y%m%d").date()

datetime.date(1990, 6, 13)

In [11]:
id = 4990061312
date = id // 100
while date > 20240000:
    date -= 10000000
datetime.strptime(str(date), "%Y%m%d").date()

datetime.date(1990, 6, 13)

In [12]:
sum(df_LF.id.isna()), sum(df_FD.id.isna())

(0, 0)

In [13]:
id - 100*(id//100)

12

# When making ghost samples, will any of our ambiguous id match between fish details and length frequencies?

In [14]:
ambiguous_FD = set(df_FD[df_FD.id>2024000000].id)
ambiguous_LF = set(df_LF[df_LF.id>2024000000].id)
ambiguous_SD = set(df_SD[df_SD.id>2024000000].id)

# YES
# we need to further disambiguate: maybe add 20/40 to months, this number would never occur naturally

In [15]:
# let's make a summary of ambiguous dates

df_A_FD = pd.DataFrame(ambiguous_FD, columns=['id_FD'])
df_A_FD['date'] = df_A_FD['id_FD'] // 100
while df_A_FD.date.max() > 20240000:
    df_A_FD.loc[df_A_FD.date>20240000, 'date'] -= 10000000
    
df_A_LF = pd.DataFrame(ambiguous_LF, columns=['id_LF'])
df_A_LF['date'] = df_A_LF['id_LF'] // 100
while df_A_LF.date.max() > 20240000:
    df_A_LF.loc[df_A_LF.date>20240000, 'date'] -= 10000000
    
df_A_SD = pd.DataFrame(ambiguous_SD, columns=['id_SD'])
df_A_SD['date'] = df_A_SD['id_SD'] // 100
while df_A_SD.date.max() > 20240000:
    df_A_SD.loc[df_A_SD.date>20240000, 'date'] -= 10000000

# group and split into columns
df_A_FD = df_A_FD.groupby('date')['id_FD'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_FD.columns = ['id_FD' for _ in range(df_A_FD.shape[1])]

df_A_LF = df_A_LF.groupby('date')['id_LF'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_LF.columns = ['id_LF' for _ in range(df_A_LF.shape[1])]

df_A_SD = df_A_SD.groupby('date')['id_SD'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_SD.columns = ['id_SD' for _ in range(df_A_SD.shape[1])]

# create a dataframe summarising ambiguous data
df_AMBIGUOUS = pd.merge(
    df_A_FD,
    pd.merge(
        df_A_SD,
        df_A_LF,
        on='date', 
        how='outer'
    ),
    on='date', 
    how='outer'
).sort_index().astype('Int64')

df_AMBIGUOUS.loc[df_AMBIGUOUS['id_FD'].any(axis=1), 'FLAG_FD_AMBIGUOUS'] = True
df_AMBIGUOUS.loc[df_AMBIGUOUS['id_LF'].any(axis=1), 'FLAG_LF_AMBIGUOUS'] = True
df_AMBIGUOUS.loc[df_AMBIGUOUS['id_SD'].any(axis=1), 'FLAG_SD_AMBIGUOUS'] = True

df_AMBIGUOUS['FLAG_AMBIGUITY_OVERLAP'] = df_AMBIGUOUS.FLAG_FD_AMBIGUOUS.notnull()*1 + df_AMBIGUOUS.FLAG_LF_AMBIGUOUS.notnull()*1 + df_AMBIGUOUS.FLAG_SD_AMBIGUOUS.notnull()*1

In [16]:
# these are our problems if we are making ghost samples

df_AMBIGUOUS.loc[df_AMBIGUOUS.FLAG_AMBIGUITY_OVERLAP > 1].dropna(axis=1).T

date                      19980501    20000607    20000609    20010516  \
id_FD                   2998050105  3000060726  3000060926  3001051605   
id_LF                   2998050105  4000060726  4000060926  3001051605   
id_LF                   3998050105  3000060726  3000060926  4001051605   
FLAG_FD_AMBIGUOUS             True        True        True        True   
FLAG_LF_AMBIGUOUS             True        True        True        True   
FLAG_AMBIGUITY_OVERLAP           2           2           2           2   

date                      20090520    20090522    20100511    20100519  \
id_FD                   3009052026  3009052226  3010051126  3010051926   
id_LF                   3009052026  3009052226  3010051126  3010051926   
id_LF                   4009052026  4009052226  4010051126  4010051926   
FLAG_FD_AMBIGUOUS             True        True        True        True   
FLAG_LF_AMBIGUOUS             True        True        True        True   
FLAG_AMBIGUITY_OVERLAP           2           2           2           2   

date                      20100527  
id_FD                   3010052726  
id_LF                   3010052726  
id_LF                   4010052726  
FLAG_FD_AMBIGUOUS             True  
FLAG_LF_AMBIGUOUS             True  
FLAG_AMBIGUITY_OVERLAP           2

# Potential Sorting Issues? LF? FD?
* Yes, can't disambiguate without making guesses. Left flagged, and will be rejected on import.
#### FLAG_MISNUMBERED_FISH_DETAILS

# More Sorting Issue Calcs and Checks
* most can be disambiguated using AM PM (cleaning sheet updated to do this)

# Trap Supervisors Linked to Sites
* check this to compare to output from dm_apps views

In [17]:
# looks correct

RELEVANT_SUPERVISORS = df_SD.groupby(['NAME', 'SITE1']).count().reset_index()[['SITE1', 'NAME']].sort_values('SITE1').reset_index(drop=True).rename({'SITE1': 'site'}, axis=1)

for site in RELEVANT_SUPERVISORS.site.unique():
    print('\nSite', site, '--', ', '.join(list(RELEVANT_SUPERVISORS.query('site == @site').NAME)))


Site 1 -- Marilyn Gillis, Charles McDaniel

Site 2 -- Richard Gillis

Site 4 -- Charles McDaniel

Site 5 -- Anthony Cameron

Site 6 -- Irene MacIsaac

Site 7 -- Eleanor MacLellan

Site 8 -- Colin Gillis

Site 9 -- John A Chisholm

Site 11 -- Darlene Cameron, Josh Fraser, Joshua Fraser

Site 12 -- Martin E Cameron

Site 15 -- Pat Chisholm, John A Chisholm

Site 17 -- Pat Chisholm, John A Chisholm

Site 18 -- Eleanor McDaniel, Brian Doyle

Site 19 -- John A Chisholm

Site 20 -- John Coady

Site 21 -- Jim Coady

Site 23 -- John A Coady

Site 24 -- James A Hirtle

Site 25 -- Robert Peters

Site 26 -- Gerard V Chiasson, Pierre Chiasson

Site 27 -- Leo Chiasson

Site 28 -- Gerard MacFarlane

Site 29 -- Donald MacLeod

Site 30 -- Robert MacLeod, Donald MacLeod

Site 31 -- Robert MacLeod

Site 32 -- Robert MacLeod, Catherine MacLeod

Site 33 -- Daniel Stewart, Michael J Stewart

Site 34 -- Joan Ingram, Fred Ingram, Ronald J Stewart

Site 35 -- John MacLellan, Jack MacLellan

Site 37 -- Floren

In [18]:
# looks correct

for supervisor in RELEVANT_SUPERVISORS.sort_values('NAME').NAME.unique():
    print(supervisor, '--', ', '.join(str(x) for x in list(RELEVANT_SUPERVISORS.query('NAME == @supervisor').site.unique())))

Alexander Gillis -- 56
Alexander MacDonald -- 39
Allan B Gillis -- 46
Anthony Cameron -- 5
Brian Doyle -- 18
Brian MacFarlane -- 66
Bruce MacLellan -- 51
Bruce McLellan -- 51
Catherine MacFarlane -- 52
Catherine MacLeod -- 32
Charles McDaniel -- 1, 4
Chris MacLean -- 68
Colin Gillis -- 8, 90
Daniel Stewart -- 33
Darlene Cameron -- 11
David MacKinnon -- 39
Donald D Gillis -- 43
Donald J Gillis -- 37
Donald M Campbell -- 45
Donald MacEachern -- 63
Donald MacLeod -- 29, 30
Donelda M Gillis -- 47
Eleanor MacLellan -- 7
Eleanor McDaniel -- 18, 91
Elizabeth MacKinnon -- 62
Elizabeth/Vincent MacKinnon -- 62
Finley MacDonald -- 62
Finley Stewart -- 44
Florence Gillis -- 37
Fred Ingram -- 34
Gerard MacFarlane -- 28
Gerard V Chiasson -- 26
Harold MacFarlane -- 67
Hugh J Gillis -- 38
Hughie MacDonnell -- 61
Irene MacIsaac -- 6
Jack MacLellan -- 35
James A Hirtle -- 24
James MacFarlane -- 55, 65
Jim Coady -- 21
Jim MacFarlane -- 62
Joan Ingram -- 34
John A Chisholm -- 9, 15, 17, 19
John A Coady --

# Comparing Comments with Provided Data

In [19]:
# unsorted df
df1SD = pd.read_pickle(r'.\intermediate_data_and_calcs\unsorted\df_SD.pickle')
df1FD = pd.read_pickle(r'.\intermediate_data_and_calcs\unsorted\df_FD.pickle')
df1LF = pd.read_pickle(r'.\intermediate_data_and_calcs\unsorted\df_LF.pickle')

dtypes_FD = dict(df_FD.dtypes)
dtypes_LF = dict(df_LF.dtypes)
dtypes_SD = dict(df_SD.dtypes)

del dtypes_FD['DATETIME']
del dtypes_LF['DATETIME']
del dtypes_SD['DATETIME']

df2FD = pd.read_csv(r'.\provided data\revised with comments\gaspereau_fish_details_AD.csv', dtype=dtypes_FD)
df2LF = pd.read_csv(r'.\provided data\revised with comments\gaspereau_length_frequencies_AD.csv', dtype=dtypes_LF)
df2SD = pd.read_csv(r'.\provided data\revised with comments\gaspereau_sample_data_AD.csv', dtype=dtypes_SD)

df2FD['DATETIME'] = pd.to_datetime(df2FD['DATETIME'])
df2LF['DATETIME'] = pd.to_datetime(df2LF['DATETIME'])
df2LF['DATETIME'] = pd.to_datetime(df2LF['DATETIME'].dt.date)  # remove time of day
df2SD['DATETIME'] = pd.to_datetime(df2SD['DATETIME'])

In [20]:
# id are different because of recent disambiguation for ghost sampels:
# drop id columns when checking for changes
(
    (df1FD.id.astype(str).str[1:] == df2FD.id.astype(str).str[1:]).all(),
    (df1LF.id.astype(str).str[1:] == df2LF.id.astype(str).str[1:]).all(), 
    (df1SD.id.astype(str).str[1:] == df2SD.id.astype(str).str[1:]).all()
)

(True, True, True)

### Fish Details

In [21]:
# cast everything to a string to compare differences
df1FD.index = df1FD.index.rename('')

# drop new columns and id column
drop_FD = [x for x in list(df1FD.columns) if x not in list(df2FD.columns)]

FD_diff = (
    df1FD.drop(drop_FD+['id'], axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
    != df2FD.drop('id', axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
)

FD_diff_columns = list()
for i, row in pd.DataFrame(FD_diff.any()).reset_index().rename({'index':'field', 0:'differences'}, axis=1).iterrows():
    if row['differences']:
        FD_diff_columns.append(row['field'])

In [22]:
FD_differences = pd.concat([
    df2FD[FD_diff][FD_diff.any(axis=1)][FD_diff_columns],
    df1FD.drop(drop_FD, axis=1)[FD_diff][FD_diff.any(axis=1)][FD_diff_columns]
], axis=1)

In [23]:
# differences all look reasonable
FD_differences.astype(str).replace({'TRUE':'True','<NA>':'','nan':''})

FL_WET FL_FROZEN    FL_STD MATURITY SITE2 FLAG_AM_PM_PERIOD FL_WET  \
14053                                         2                            
14054                                         2                            
14055                                         2                            
14056                                         2                            
14057                                         2                            
14058                                         2                            
14059                                         2                            
14060                                         2                            
14061                                         2                            
14062                                         2                            
14063                                         2                            
14064                                         2                            
14065                                         2                            
14066                                         2                            
14067                                         2                            
14068                                         2                            
14069                                         2                            
14070                                         2                            
14071                                         2                            
14072                                         2                            
14073                                         2                            
14074                                         2                            
14075                                         2                            
14076                                         2                            
14077                                         2                            
14078                                         2                            
14079                                         2                            
14080                                         2                            
14081                                         2                            
14082                                         2                            
14083                                         2                            
14084                                         2                            
14085                                         2                            
14086                                         2                            
14087                                         2                            
14088                                         2                            
15925                                                       AM/PM          
15926                                                       AM/PM          
15927                                                       AM/PM          
15928                                                       AM/PM          
15929                                                       AM/PM          
15930                                                       AM/PM          
15931                                                       AM/PM          
15932                                                       AM/PM          
15933                                                       AM/PM          
15934                                                       AM/PM          
15935                                                       AM/PM          
15936                                                       AM/PM          
15937                                                       AM/PM          
15938                                                       AM/PM          
15939                                                       AM/PM          
15940                                                 

### Length Frequencies

In [24]:
LF_diff = (
    df1LF.drop('id', axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
    != df2LF.drop('id', axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
)

LF_diff_columns = list()
for i, row in pd.DataFrame(LF_diff.any()).reset_index().rename({'index':'field', 0:'differences'}, axis=1).iterrows():
    if row['differences']:
        LF_diff_columns.append(row['field'])

In [25]:
LF_differences = pd.concat([
    df2LF[LF_diff][LF_diff.any(axis=1)][LF_diff_columns],
    df1LF[LF_diff][LF_diff.any(axis=1)][LF_diff_columns]
], axis=1)

In [26]:
# Perfect. Already got this one.
LF_differences.astype(str).replace({'TRUE':'True','<NA>':'','nan':''})

lgth  Flbin length_bin_id   lgth  Flbin length_bin_id
10851  235.0  235.0          23.5  265.0  265.0          26.5

### Samples

In [27]:
[x for x in list(df2SD.columns) if x not in list(df1SD.columns)]

['period']

In [28]:
[x for x in list(df1SD.columns) if x not in list(df2SD.columns)]

['AM_PM_PERIOD']

In [29]:
# slightly better merge with newer version (AM_PM_PERIOD)
# => drop these columns
pd.concat([
    df1SD['AM_PM_PERIOD'].fillna(0), 
    df2SD['period'].fillna(0)], 
    axis=1
)[(df1SD['AM_PM_PERIOD'].fillna(0) != df2SD['period'].fillna(0))]

AM_PM_PERIOD period
3077            AM      0
3090            PM      0
3102            PM      0
3110            PM      0
3119            AM      0
3121            PM      0
3122            AM      0
3124            PM      0
3135            AM      0
3144            PM      0
3157            AM      0
3171            AM      0
3175            PM      0
3204            PM      0
3210            PM      0
3234            PM      0
3237            AM      0
3246            AM      0
3262            PM      0
3267            AM      0
3270            PM      0
3305            AM      0
3306            AM      0
3327            PM      0
3330            AM      0
3350            PM      0
3356            AM      0
3361            AM      0
3368            AM      0
3374            AM      0
3375            AM      0
3379            AM      0
3381            PM      0
3382            PM      0
3392            AM      0
3402            AM      0
3419            PM      0
3437            AM      0
3754            PM      0
5130            AM     PM
5257            AM      0
5347            AM      0
5485            AM      0
5651            AM      0
5653            PM      0
5671            AM      0
5823            AM      0
6132            PM      0
6551            AM      0
7288            AM      0
7909            PM      0
8200            AM     PM
8330            AM      0
8422            PM      0
8940            AM      0
9381            PM      0
10211           AM     PM
10370           PM      0
11069           AM      0
11789           PM      0
11942           AM      0
12005           AM      0
13373           AM     PM
13452           PM      0
14740           PM      0

In [30]:
# cast everything to a string to compare differences
df1SD.index = df1SD.index.rename('')

# drop new columns and id column
drop_SD = [x for x in list(df1SD.columns) if x not in list(df2SD.columns)]

SD_diff = (
    df1SD.drop(['id', 'AM_PM_PERIOD'], axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
    != df2SD.drop(['id', 'period'], axis=1).astype(str).replace({'TRUE':'True','NaN':'','<NA>':'','nan':''})
)

SD_diff_columns = list()
for i, row in pd.DataFrame(SD_diff.any()).reset_index().rename({'index':'field', 0:'differences'}, axis=1).iterrows():
    if row['differences']:
        SD_diff_columns.append(row['field'])

In [31]:
SD_differences = pd.concat([
    df2SD[SD_diff][SD_diff.any(axis=1)][SD_diff_columns],
    df1SD[SD_diff][SD_diff.any(axis=1)][SD_diff_columns]
], axis=1)

In [32]:
# nothing new here
SD_differences.astype(str).replace({'TRUE':'True','<NA>':'','nan':''})

total_fish_measured wt_lbs total_fish_measured wt_lbs
14740                                          259.0   92.0

In [33]:
SD_diff.any()

DIST                               False
RIVER                              False
NAME                               False
code                               False
GEAR                               False
SITE_NO                            False
no_nets                            False
YEAR                               False
MM                                 False
DD                                 False
Week                               False
catch_lbs                          False
catch_kg                           False
hours_fished                       False
zone                               False
last_name                          False
comments                           False
bycatch_sbass                      False
bycatch_shad                       False
bycatch_other                      False
DATETIME                           False
SITE1                              False
SITE2                              False
remarks                            False
total_fish_prese

In [34]:
# only this one - slightly better merge with LF
pd.concat([
    df1SD.iloc[14740],
    df2SD.iloc[14740]
], axis=1).T

DIST   RIVER             NAME code GEAR SITE_NO no_nets  YEAR MM  DD  \
14740    2  SWMARG  Kevin MacKinnon  NaN   81      64       1  2018  6  13   
14740    2  SWMARG  Kevin MacKinnon  NaN   81      64       1  2018  6  13   

      Week catch_lbs catch_kg hours_fished   zone  last_name comments  \
14740    7   270.000  122.500          4.5  upper  Mackinnon      NaN   
14740    7   270.000  122.500          4.5  upper  Mackinnon      NaN   

      bycatch_sbass bycatch_shad bycatch_other   DATETIME SITE1 SITE2 remarks  \
14740           NaN          NaN           NaN 2018-06-13    64  <NA>     NaN   
14740           NaN          NaN           NaN 2018-06-13    64   NaN     NaN   

               id total_fish_preserved total_fish_measured AM_PM_PERIOD  \
14740  2018061364               29.000             259.000           PM   
14740  2018061364               29.000                 NaN          NaN   

      wt_lbs FLAG_DATETIME FLAG_HOURS_FISHED FLAG_SITE  \
14740 92.000           NaN               NaN       NaN   
14740    NaN           NaN               NaN       NaN   

      FLAG_AM_PM_PERIOD_DISCREPANCIES period  
14740                             NaN    NaN  
14740                             NaN    NaN

# COMMENT re Fish Details FLAG_SITE:

    CONVERT E.MACFARLANE, Eric Mac, Eric MacFarlane, J McFarlane to John Eric MacFarlane.

    JA Coady keep as is

    Multiple sites, keep as is. 

    I see 118 samples [fish details] from 1989 that are associated with ‘blank’. Can you convert ‘blank’ to Unknown and upload.

In [35]:
# this is from the next step (out of order)
df_SD_cleaned = pd.read_pickle('df_SD_cleaned.pickle')

In [36]:
list(df_FD[df_FD.SITE.isna()].id.unique()), df_FD[df_FD.SITE.isna()].shape[0]

([1989060100, 1989060200, 1989060700], 118)

In [37]:
# Yes, these are imported as ghost samples
df_SD_cleaned[df_SD_cleaned.id.isin(list(df_FD[df_FD.SITE.isna()].id.unique()))].remarks

15427    SITE_notes: nan; GHOST SAMPLE: created to match with unmatched Fish Details and/or Length Frequencies
15428    SITE_notes: nan; GHOST SAMPLE: created to match with unmatched Fish Details and/or Length Frequencies
15430    SITE_notes: nan; GHOST SAMPLE: created to match with unmatched Fish Details and/or Length Frequencies
Name: remarks, dtype: object

### let's make a list of all slightly dubious matches (inexact sites)

In [38]:
# FD ambiguous site names
FD_sites_id = df_FD[(~df_FD['SITE'].fillna(0).astype(str).str.isnumeric()) & (df_FD.SITE!='1A') & (df_FD.SITE!='1B')].id.unique()

# here are our troublesome sites
df_FD[df_FD.id.isin(FD_sites_id)].groupby('id').first()[['DATETIME', 'SITE', 'SITE1']]

DATETIME             SITE  SITE1
id                                           
1993052060 1993-05-20           60, 52     60
1993052301 1993-05-23             1, 8      1
1993052437 1993-05-24         37,38,35     37
1993052660 1993-05-26            60,52     60
1993052701 1993-05-27              1,8      1
1993060305 1993-06-03              5,8      5
1994051012 1994-05-10            12,17     12
1994051115 1994-05-11            15,17     15
1994051133 1994-05-11         33,48,49     33
1994051411 1994-05-14            11,12     11
1994051605 1994-05-16          5,11,17      5
1994051633 1994-05-16         33,48,49     33
1994051737 1994-05-17            37,60     37
1994052352 1994-05-23            52,60     52
1994052538 1994-05-25            38,52     38
1994052701 1994-05-27           1,8,26      1
1994052733 1994-05-27         33,41,60     33
1994052801 1994-05-28              1,8      1
1994053001 1994-05-30              1,8      1
1994053101 1994-05-31              1,8      1
1994053135 1994-05-31            35,60     35
1994060735 1994-06-07            35,37     35
1995052441 1995-05-24            41,49     41
1995052552 1995-05-25         52,60,38     52
1995053160 1995-05-31         60,52,35     60
1995060241 1995-06-02            41,35     41
1995060301 1995-06-03            01,02      1
1995060560 1995-06-05            60,41     60
1995060635 1995-06-06            35,41     35
1995060938 1995-06-09         38,60,52     38
1995061149 1995-06-11            49,41     49
1995061235 1995-06-12            35,52     35
1996052212 1996-05-22            12,26     12
1996060802 1996-06-08              2,5      2
1997052205 1997-05-22             5,17      5
2001060615 2001-06-06            15,17     15
2002050915 2002-05-09            15,17     15
2002051715 2002-05-17            15,17     15
2002052905 2002-05-29             5,17      5
2003051625 2003-05-16            25,26     25
2016053195 2016-05-31   11K3 (435,211)     95
2016061495 2016-06-14   11K3 (435,211)     95
2016062195 2016-06-21   11K3 (435,211)     95
2017053092 2017-05-30     E.MACFARLANE     92
2017060692 2017-06-06     E.MACFARLANE     92
2018052992 2018-05-29  Eric MacFarlane     92
2018060892 2018-06-08  Eric MacFarlane     92
2018062792 2018-06-27  Eric MacFarlane     92
2019060793 2019-06-07      J McFarlane     93
2019060794 2019-06-07         JA Coady     94
2019061493 2019-06-14      J McFarlane     93
2019062594 2019-06-25         JA Coady     94
2021052092 2021-05-20         Eric Mac     92
2021060192 2021-06-01         Eric Mac     92
2021060392 2021-06-03         Eric Mac     92
2021061592 2021-06-15         Eric Mac     92
2021062392 2021-06-23         Eric Mac     92
3001051605 2001-05-16             5,26      5
4014053047 2014-05-30         47 or 62     47

In [39]:
# all flagged for import into dm_apps
df_SD_cleaned[df_SD_cleaned.id.isin(FD_sites_id)][['id', 'FLAG_GHOST_SAMPLE', 'FLAG_AMBIGUOUS_MATCH', 'FLAG_AMBIGUOUS_SAMPLE']]

id FLAG_GHOST_SAMPLE FLAG_AMBIGUOUS_MATCH FLAG_AMBIGUOUS_SAMPLE
5213   1993052060               NaN                 True                   NaN
5245   1993052301               NaN                 True                   NaN
5272   1993052437               NaN                 True                   NaN
5310   1993052660               NaN                 True                   NaN
5311   1993052701               NaN                 True                   NaN
5423   1993060305               NaN                 True                   NaN
5569   1994051012               NaN                 True                   NaN
5598   1994051411               NaN                 True                   NaN
5610   1994051605               NaN                 True                   NaN
5632   1994051737               NaN                 True                   NaN
5727   1994052352               NaN                 True                   NaN
5778   1994052801               NaN                 True                   NaN
5804   1994053001               NaN                 True                   NaN
5818   1994053101               NaN                 True                   NaN
6200   1995061149               NaN                 True                   NaN
6293   1996052212               NaN                 True                   NaN
6366   1997052205               NaN                 True                   NaN
8136   2001060615               NaN                 True                   NaN
8301   2002050915               NaN                 True                   NaN
8384   2002051715               NaN                 True                   NaN
8488   2002052905               NaN                 True                   NaN
8716   2003051625               NaN                 True                   NaN
15266  1995060301              True                  NaN                   NaN
15279  1995052441              True                  NaN                   NaN
15288  1995060635              True                  NaN                   NaN
15290  1995060938              True                  NaN                   NaN
15293  1995061235              True                  NaN                   NaN
15301  1996060802              True                  NaN                   NaN
15311  3001051605              True                  NaN                  True
15354  2017053092              True                  NaN                   NaN
15355  2017060692              True                  NaN                   NaN
15372  1994051115              True                  NaN                   NaN
15373  1994060735              True                  NaN                   NaN
15375  1994052538              True                  NaN                   NaN
15393  2019060794              True                  NaN                   NaN
15394  2019060793              True                  NaN                   NaN
15395  2019061493              True                  NaN                   NaN
15396  2019062594              True                  NaN                   NaN
15437  1994051133              True                  NaN                   NaN
15438  1994051633              True                  NaN                   NaN
15439  1994052701              True                  NaN                   NaN
15440  1994052733              True                  NaN                   NaN
15441  1994053135              True                  NaN                   NaN
15443  1995052552              True                  NaN                   NaN
15445  1995053160              True                  NaN                   NaN
15446  1995060241              True                  NaN                   NaN
15447  1995060560              True                  NaN                   NaN
15457  2016053195              True                  NaN                   NaN
15458  2016061495              True                  NaN                   NaN
15459  2016062195              True                  NaN         

In [40]:
# LF ambiguous site names
LF_sites_id = df_LF[(~df_LF['site'].fillna(0).astype(str).str.isnumeric()) & (df_LF.site!='1A') & (df_LF.site!='1B')].id.unique()
df_LF[df_LF.id.isin(LF_sites_id)].groupby('id').first()[['DATETIME', 'site', 'SITE1']]

# list of the distinct sites that are ambiguous
# df_LF[df_LF.id.isin(LF_sites_id)].groupby('id').first()[['DATETIME', 'site', 'SITE1']].replace({'Jimmy MacFarlane':'Eric McFarlane', 93:92}).groupby(['site', 'SITE1']).count().reset_index().drop('DATETIME', axis=1).T

DATETIME               site  SITE1
id                                             
1991060129 1991-06-01              29,30     29
1993051117 1993-05-11              17,33     17
1993052301 1993-05-23                1,8      1
1993052435 1993-05-24           35,37,38     35
1993052701 1993-05-27                1,8      1
1993060305 1993-06-03                5,8      5
1993060352 1993-06-03              52,60     52
1994051012 1994-05-10              12,17     12
1994051115 1994-05-11              15,17     15
1994051149 1994-05-11              49,33     49
1994051438 1994-05-14              38,41     38
1994051617 1994-05-16           17,12, 5     17
1994051648 1994-05-16          48,38, 37     48
1994051737 1994-05-17              37,60     37
1994052360 1994-05-23              60,52     60
1994052538 1994-05-25              38,52     38
1994052617 1994-05-26               17,5     17
1994052726 1994-05-27               26,1     26
1994052741 1994-05-27          41,60, 33     41
1994052801 1994-05-28                1,8      1
1994053001 1994-05-30                1,8      1
1994053101 1994-05-31                1,8      1
1994053160 1994-05-31              60,35     60
1994060735 1994-06-07              35,37     35
1995052205 1995-05-22               5,12      5
1995052441 1995-05-24              41,49     41
1995052538 1995-05-25           38,52,60     38
1995052902 1995-05-29                2,5      2
1995053017 1995-05-30              17,26     17
1995053138 1995-05-31           38,52,60     38
1995060235 1995-06-02              35,41     35
1995060301 1995-06-03                1,2      1
1995060541 1995-06-05           41,52,60     41
1995060635 1995-06-06              35,41     35
1995060938 1995-06-09           38,52,60     38
1995061141 1995-06-11              41,49     41
1995061235 1995-06-12              35,52     35
1996052212 1996-05-22              12,26     12
1996060802 1996-06-08                2,5      2
1997052205 1997-05-22               5,17      5
2001060615 2001-06-06              15,17     15
2002052905 2002-05-29               5,17      5
2003051625 2003-05-16              25,26     25
2017053092 2017-05-30     Eric McFarlane     92
2017060692 2017-06-06     Eric McFarlane     92
2019060793 2019-06-07   Jimmy MacFarlane     93
2019060794 2019-06-07  John Albert Coady     94
2019061493 2019-06-14   Jimmy MacFarlane     93
2019062594 2019-06-25  John Albert Coady     94
3001051605 2001-05-16               5,26      5

In [41]:
# all flagged for import into dm_apps
df_SD_cleaned[df_SD_cleaned.id.isin(LF_sites_id)][['id', 'FLAG_GHOST_SAMPLE', 'FLAG_AMBIGUOUS_MATCH', 'FLAG_AMBIGUOUS_SAMPLE']]

id FLAG_GHOST_SAMPLE FLAG_AMBIGUOUS_MATCH FLAG_AMBIGUOUS_SAMPLE
5245   1993052301               NaN                 True                   NaN
5311   1993052701               NaN                 True                   NaN
5423   1993060305               NaN                 True                   NaN
5435   1993060352               NaN                 True                   NaN
5569   1994051012               NaN                 True                   NaN
5615   1994051617               NaN                 True                   NaN
5621   1994051648               NaN                 True                   NaN
5632   1994051737               NaN                 True                   NaN
5728   1994052360               NaN                 True                   NaN
5757   1994052617               NaN                 True                   NaN
5771   1994052726               NaN                 True                   NaN
5774   1994052741               NaN                 True                   NaN
5778   1994052801               NaN                 True                   NaN
5804   1994053001               NaN                 True                   NaN
5818   1994053101               NaN                 True                   NaN
5831   1994053160               NaN                 True                   NaN
6020   1995052205               NaN                 True                   NaN
6293   1996052212               NaN                 True                   NaN
6366   1997052205               NaN                 True                   NaN
8136   2001060615               NaN                 True                   NaN
8488   2002052905               NaN                 True                   NaN
8716   2003051625               NaN                 True                   NaN
15226  1991060129              True                  NaN                   NaN
15239  1993052435              True                  NaN                   NaN
15240  1993051117              True                  NaN                   NaN
15261  1995052902              True                 True                   NaN
15262  1995053017              True                  NaN                   NaN
15266  1995060301              True                  NaN                   NaN
15279  1995052441              True                  NaN                   NaN
15280  1995052538              True                  NaN                   NaN
15284  1995053138              True                  NaN                   NaN
15285  1995060235              True                  NaN                   NaN
15287  1995060541              True                  NaN                   NaN
15288  1995060635              True                  NaN                   NaN
15290  1995060938              True                  NaN                   NaN
15292  1995061141              True                  NaN                   NaN
15293  1995061235              True                  NaN                   NaN
15301  1996060802              True                  NaN                   NaN
15311  3001051605              True                  NaN                  True
15354  2017053092              True                  NaN                   NaN
15355  2017060692              True                  NaN                   NaN
15372  1994051115              True                  NaN                   NaN
15373  1994060735              True                  NaN                   NaN
15374  1994051438              True                  NaN                   NaN
15375  1994052538              True                  NaN                   NaN
15376  1994051149              True                  NaN                   NaN
15393  2019060794              True                  NaN                   NaN
15394  2019060793              True                  NaN                   NaN
15395  2019061493              True                  NaN                   NaN
15396  2019062594              True                  NaN         

In [42]:
# SD
SD_sites_id = df_SD[(~df_SD['SITE_NO'].fillna(0).astype(str).str.isnumeric()) & (df_SD.SITE_NO!='1A') & (df_SD.SITE_NO!='1B')].id.unique()

In [43]:
# how many dubious sites are there?
dubious_sites_id = sorted([x for x in set(list(LF_sites_id)+list(FD_sites_id)+list(SD_sites_id))])
len(dubious_sites_id)

103

In [44]:
# all flagged for import into dm_apps
df_SD_cleaned[df_SD_cleaned.id.isin(dubious_sites_id)][['FLAG_GHOST_SAMPLE', 'FLAG_AMBIGUOUS_MATCH', 'FLAG_AMBIGUOUS_SAMPLE']].any(axis=1).all()

True

# Create Dubious Matches Dataframe For Notes and Flags for Import

In [45]:
# merge all df together with inconsistent sites
# see if the inconsistencies are dubious or obvious

dubious_matches_SD_LF_FD = pd.merge(
    pd.merge(
        df_SD[df_SD.id.isin(dubious_sites_id)][['DATETIME', 'id', 'SITE_NO', 'FLAG_SITE']].rename({'FLAG_SITE':'FLAG_SITE_SD'}, axis=1),
        df_FD[df_FD.id.isin(dubious_sites_id)][[
            'id', 'SITE', 'PERIOD', 'FLAG_SITE', 'FLAG_MULTIPLE_SAMPLE_POSSIBILITIES', 'FLAG_NO_MATCHING_SAMPLE'
        ]].rename({'FLAG_SITE':'FLAG_SITE_FD', 'FLAG_NO_MATCHING_SAMPLE':'FLAG_NO_MATCHING_SAMPLE_FD'}, axis=1).groupby('id').first().reset_index(),
        on='id',
        how='left'  # only care about 3 way merge that include samples
    ).sort_values('id'),
    df_LF[df_LF.id.isin(dubious_sites_id)][[
        'id', 'site', 'period', 'FLAG_SITE', 'FLAG_NO_MATCHING_SAMPLE'
    ]].rename({'FLAG_SITE':'FLAG_SITE_LF', 'FLAG_NO_MATCHING_SAMPLE':'FLAG_NO_MATCHING_SAMPLE_LF'}, axis=1).groupby('id').first().reset_index(),
    on='id',
    how='left'
).sort_values('id')[[
    'DATETIME', 'id', 'SITE_NO', 'site', 'SITE', 'period', 'PERIOD',
    'FLAG_SITE_SD', 'FLAG_SITE_FD', 'FLAG_MULTIPLE_SAMPLE_POSSIBILITIES',
    'FLAG_NO_MATCHING_SAMPLE_FD', 'FLAG_SITE_LF', 'FLAG_NO_MATCHING_SAMPLE_LF'
]]

# dubious_matches_SD_LF_FD

In [46]:
# flag these and add a remark in SD about the differences in sites between tables
dubious_matches_SD_LF_FD['SITE_AMBIGUITIES'] = (
    'SITE AMBIGUITIES: Samples - ' + dubious_matches_SD_LF_FD['SITE_NO'].astype(str) 
     + '; Fish details - ' + dubious_matches_SD_LF_FD['SITE'].astype(str) 
     + '; Length frequencies - ' + dubious_matches_SD_LF_FD['site'].astype(str)
)

In [47]:
# no need to note period - these all match
dubious_matches_SD_LF_FD[(dubious_matches_SD_LF_FD.period != dubious_matches_SD_LF_FD.PERIOD) & dubious_matches_SD_LF_FD.PERIOD.notnull() & dubious_matches_SD_LF_FD.period.notnull()]

Empty DataFrame
Columns: [DATETIME, id, SITE_NO, site, SITE, period, PERIOD, FLAG_SITE_SD, FLAG_SITE_FD, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_NO_MATCHING_SAMPLE_FD, FLAG_SITE_LF, FLAG_NO_MATCHING_SAMPLE_LF, SITE_AMBIGUITIES]
Index: []

In [49]:
dubious_matches_SD_LF_FD[['DATETIME', 'id', 'SITE_NO', 'site', 'SITE', 'period', 'PERIOD', 'SITE_AMBIGUITIES']]

DATETIME          id SITE_NO       site      SITE period PERIOD  \
0  1993-05-20  1993052060      60         60    60, 52     AM     AM   
1  1993-05-23  1993052301       1        1,8      1, 8     PM     PM   
2  1993-05-24  1993052437      37        NaN  37,38,35    NaN     AM   
3  1993-05-26  1993052660      60         60     60,52     PM     PM   
4  1993-05-27  1993052701       1        1,8       1,8     PM     PM   
5  1993-06-03  1993060305       5        5,8       5,8     AM     AM   
6  1993-06-03  1993060352      52      52,60       NaN     PM    NaN   
7  1994-05-10  1994051012      12      12,17     12,17     PM     PM   
8  1994-05-14  1994051411      11        NaN     11,12    NaN     AM   
9  1994-05-16  1994051605       5        NaN   5,11,17    NaN     AM   
10 1994-05-16  1994051617      17   17,12, 5       NaN     AM    NaN   
11 1994-05-16  1994051648      48  48,38, 37       NaN     AM    NaN   
12 1994-05-17  1994051737      37      37,60     37,60     PM     PM   
13 1994-05-23  1994052352      52        NaN     52,60    NaN     AM   
14 1994-05-23  1994052360      60      60,52       NaN     AM    NaN   
15 1994-05-26  1994052617      17       17,5        17     AM     AM   
16 1994-05-27  1994052726      26       26,1       NaN     PM    NaN   
17 1994-05-27  1994052741      41  41,60, 33       NaN     AM    NaN   
18 1994-05-28  1994052801       1        1,8       1,8     PM     PM   
19 1994-05-30  1994053001       1        1,8       1,8     AM     AM   
20 1994-05-31  1994053101       1        1,8       1,8     PM     PM   
21 1994-05-31  1994053160      60      60,35       NaN     AM    NaN   
22 1995-05-21  1995052190    1A,8        NaN       NaN    NaN    NaN   
23 1995-05-22  1995052205       5       5,12       NaN     AM    NaN   
24 1995-05-22  1995052290    1A,8        NaN       NaN    NaN    NaN   
25 1995-05-23  1995052390    1A,8        NaN       NaN    NaN    NaN   
26 1995-05-24  1995052490    1A,8        NaN       NaN    NaN    NaN   
27 1995-05-25  1995052590    1A,8        NaN       NaN    NaN    NaN   
28 1995-05-26  1995052690    1A,8        NaN       NaN    NaN    NaN   
29 1995-05-27  1995052790    1A,8        NaN       NaN    NaN    NaN   
30 1995-05-28  1995052890    1A,8        NaN       NaN    NaN    NaN   
31 1995-05-29  1995052990    1A,8        NaN       NaN    NaN    NaN   
32 1995-05-30  1995053090    1A,8        NaN       NaN    NaN    NaN   
33 1995-05-31  1995053190    1A,8        NaN       NaN    NaN    NaN   
34 1995-06-01  1995060190    1A,8        NaN       NaN    NaN    NaN   
35 1995-06-02  1995060290    1A,8        NaN       NaN    NaN    NaN   
36 1995-06-03  1995060390    1A,8        NaN       NaN    NaN    NaN   
37 1995-06-04  1995060490    1A,8        NaN       NaN    NaN    NaN   
38 1995-06-05  1995060590    1A,8        NaN       NaN    NaN    NaN   
39 1995-06-06  1995060690    1A,8        NaN       NaN    NaN    NaN   
40 1995-06-07  1995060790    1A,8        NaN       NaN    NaN    NaN   
41 1995-06-09  1995060990    1A,8        NaN       NaN    NaN    NaN   
42 1995-06-10  1995061090    1A,8        NaN       NaN    NaN    NaN   
43 1995-06-11  1995061149      49        NaN     49,41    NaN     PM   
44 1995-06-11  1995061190    1A,8        NaN       NaN    NaN    NaN   
45 1995-06-12  1995061290    1A,8        NaN       NaN    NaN    NaN   
46 1995-06-13  1995061390    1A,8        NaN       NaN    NaN    NaN   
47 1996-05-22  1996052212      12      12,26     12,26     AM     AM   
48 1997-05-22  1997052205       5       5,17      5,17     PM     PM   
49 2001-06-06  2001060615      15      15,17     15,17     PM     PM   
50 2002-05-09  2002050915      15        NaN     15,17    NaN     AM   
51 2002-05-17  2002051715      15        NaN     15,17    NaN     PM   
52 2002-05-29  2002052905       5       5,17      5,17     PM     PM   
53 2003-05-16  2003051625      25      25,26     25,26     AM     AM   

                                            

In [50]:
if save_dubious_details_file := False:
    dubious_matches_SD_LF_FD[['DATETIME', 'id', 'SITE_NO', 'site', 'SITE', 'period', 'PERIOD', 'SITE_AMBIGUITIES']].to_pickle(r'.\intermediate_data_and_calcs\df_ambiguous_details.pickle')
dubious_matches_SD_LF_FD = dubious_matches_SD_LF_FD[['id', 'SITE_AMBIGUITIES']]

##### now check ambiguous matches with LF and FD

In [51]:
# see if the inconsistencies are dubious or obvious

dubious_matches_LF_FD = pd.merge(
    df_FD[df_FD.id.isin(dubious_sites_id)][[
        'id', 'SITE', 'PERIOD', 'FLAG_SITE', 'FLAG_MULTIPLE_SAMPLE_POSSIBILITIES', 'FLAG_NO_MATCHING_SAMPLE'
    ]].rename({'FLAG_SITE':'FLAG_SITE_FD', 'FLAG_NO_MATCHING_SAMPLE':'FLAG_NO_MATCHING_SAMPLE_FD'}, axis=1).groupby('id').first().reset_index(),
    df_LF[df_LF.id.isin(dubious_sites_id)][[
        'id', 'site', 'period', 'FLAG_SITE', 'FLAG_NO_MATCHING_SAMPLE'
    ]].rename({'FLAG_SITE':'FLAG_SITE_LF', 'FLAG_NO_MATCHING_SAMPLE':'FLAG_NO_MATCHING_SAMPLE_LF'}, axis=1).groupby('id').first().reset_index(),
    on='id',
    how='inner'
).sort_values('id')[['id', 'site', 'SITE', 'period', 'PERIOD']]

dubious_matches_LF_FD

id               site          SITE period PERIOD
0   1993052060                 60        60, 52     AM     AM
1   1993052301                1,8          1, 8     PM     PM
2   1993052660                 60         60,52     PM     PM
3   1993052701                1,8           1,8     PM     PM
4   1993060305                5,8           5,8     AM     AM
5   1994051012              12,17         12,17     PM     PM
6   1994051115              15,17         15,17     PM     PM
7   1994051737              37,60         37,60     PM     PM
8   1994052538              38,52         38,52     PM     PM
9   1994052617               17,5            17     AM     AM
10  1994052801                1,8           1,8     PM     PM
11  1994053001                1,8           1,8     AM     AM
12  1994053101                1,8           1,8     PM     PM
13  1994060735              35,37         35,37     PM     PM
14  1995052441              41,49         41,49     AM     AM
15  1995052902                2,5             2     AM     AM
16  1995060301                1,2         01,02     AM     AM
17  1995060635              35,41         35,41     AM     AM
18  1995060938           38,52,60      38,60,52     AM     AM
19  1995061235              35,52         35,52     PM     PM
20  1996052212              12,26         12,26     AM     AM
21  1996060802                2,5           2,5     AM     AM
22  1997052205               5,17          5,17     PM     PM
23  2001060615              15,17         15,17     PM     PM
24  2002052905               5,17          5,17     PM     PM
25  2003051625              25,26         25,26     AM     AM
26  2017053092     Eric McFarlane  E.MACFARLANE     PM     PM
27  2017060692     Eric McFarlane  E.MACFARLANE     AM     AM
28  2019060793   Jimmy MacFarlane   J McFarlane     AM     AM
29  2019060794  John Albert Coady      JA Coady     AM     AM
30  2019061493   Jimmy MacFarlane   J McFarlane     PM   None
31  2019062594  John Albert Coady      JA Coady     AM   None
32  3001051605               5,26          5,26     AM     AM

In [52]:
dubious_matches_LF_FD_id = [1993052060, 1993052660, 1994052617, 1995052902]

# now remove duplicates from SD
dubious_matches_LF_FD = dubious_matches_LF_FD[dubious_matches_LF_FD.id.isin(
    [x for x in dubious_matches_LF_FD_id if x not in list(dubious_matches_SD_LF_FD.id)]
)].copy()

In [53]:
dubious_matches_LF_FD['SITE_AMBIGUITIES'] = (
    'SITE AMBIGUITIES: Fish details - ' + dubious_matches_LF_FD['SITE'].astype(str) 
     + '; Length frequencies - ' + dubious_matches_LF_FD['site'].astype(str)
)

In [54]:
dubious_matches_LF_FD = dubious_matches_LF_FD[['id', 'SITE_AMBIGUITIES']]

# SAVE NOTES TO ADD BEFORE IMPORT

In [55]:
df_ambiguous_match_notes = pd.concat([
    dubious_matches_SD_LF_FD,
    dubious_matches_LF_FD
]).sort_values('id').reset_index(drop=True)

In [56]:
if save_ambiguous_notes_pickle := False:
    df_ambiguous_match_notes.to_pickle(r'.\intermediate_data_and_calcs\df_ambiguous_match_notes.pickle')

In [57]:
# none of these will end up being disambiguated - good
df_ambiguous_match_notes.id.max()

2003051625